File: iOS/KajHobe/Managers/NotificationLocalState.swift
import Foundation
import Combine

/// Device-local read / cleared state for notifications.
///
/// Notifications are treated as transient: the SERVER is the source of truth for
/// *what* notifications exist (job_interests, notifications) and for real actions
/// (accept / reject), while THIS store — persisted in UserDefaults, scoped per user —
/// owns whether each notification is read or cleared.
///
/// Unread rule: `created_at > baseline && id ∉ readIDs && id ∉ clearedIDs`.
/// The `baseline` is stamped the first time a user is configured on a build that has
/// this store, so the entire pre-existing backlog (e.g. the historical 206) is
/// implicitly "read" and never inflates the badge — without any server migration.
///
/// Trade-off (intentional, per product decision): state lives on-device, so it does
/// not sync across devices and resets on reinstall.
///
/// Access is expected on the main thread (SwiftUI views + main-hopped callers).
final class NotificationLocalState: ObservableObject {
    static let shared = NotificationLocalState()

    /// Bumped on every mutation so SwiftUI views observing this object re-render.
    @Published private(set) var revision: Int = 0

    private var userId: String = ""
    private var readIDs: Set<String> = []
    private var clearedIDs: Set<String> = []
    private var baseline: Date = .distantPast

    private let defaults = UserDefaults.standard

    private init() {}

    // MARK: - Configuration

    /// Scope the store to a user and load their persisted state. Idempotent.
    func configure(userId: String) {
        guard !userId.isEmpty, userId != self.userId else { return }
        self.userId = userId
        load()
        revision += 1
    }

    private func key(_ suffix: String) -> String { "notif.\(suffix).\(userId)" }

    private func load() {
        readIDs = Set(defaults.stringArray(forKey: key("read")) ?? [])
        clearedIDs = Set(defaults.stringArray(forKey: key("cleared")) ?? [])

        let baselineKey = key("baseline")
        if defaults.object(forKey: baselineKey) != nil {
            baseline = Date(timeIntervalSince1970: defaults.double(forKey: baselineKey))
        } else {
            // First launch for this user on a build with local state: everything that
            // already exists is considered read. Only notifications created after now
            // can become "unread".
            baseline = Date()
            defaults.set(baseline.timeIntervalSince1970, forKey: baselineKey)
        }
    }

    private func persistRead() { defaults.set(Array(readIDs), forKey: key("read")) }
    private func persistCleared() { defaults.set(Array(clearedIDs), forKey: key("cleared")) }

    // MARK: - Queries

    func isCleared(_ id: String) -> Bool { clearedIDs.contains(id) }

    func isUnread(id: String, createdAt: Date?) -> Bool {
        if clearedIDs.contains(id) || readIDs.contains(id) { return false }
        guard let createdAt else { return false }
        return createdAt > baseline
    }

    func isUnread(id: String, createdAtISO iso: String) -> Bool {
        isUnread(id: id, createdAt: Self.date(from: iso))
    }

    // MARK: - Mutations

    func markRead(_ id: String) {
        guard !readIDs.contains(id) else { return }
        readIDs.insert(id)
        persistRead()
        revision += 1
    }

    func markRead(_ ids: [String]) {
        let newOnes = ids.filter { !readIDs.contains($0) }
        guard !newOnes.isEmpty else { return }
        readIDs.formUnion(newOnes)
        persistRead()
        revision += 1
    }

    func clear(_ id: String) {
        clearedIDs.insert(id)
        readIDs.remove(id)
        persistCleared()
        persistRead()
        revision += 1
    }

    func clear(_ ids: [String]) {
        guard !ids.isEmpty else { return }
        clearedIDs.formUnion(ids)
        readIDs.subtract(ids)
        persistCleared()
        persistRead()
        revision += 1
    }

    // MARK: - Helpers

    private static let isoFractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()
    private static let isoPlain = ISO8601DateFormatter()

    static func date(from iso: String) -> Date? {
        isoFractional.date(from: iso) ?? isoPlain.date(from: iso)
    }
}

File: Android/app/src/main/java/com/kajhobe/app/data/local/NotificationLocalState.kt
package com.kajhobe.app.data.local

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringSetPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.kajhobe.app.data.model.parseIsoMillis
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

private val Context.notifDataStore by preferencesDataStore("notification_local_state")

/**
 * Device-local read / cleared state for notifications — the Android port of iOS
 * `NotificationLocalState`.
 *
 * Notifications are transient: the SERVER is the source of truth for *what* exists
 * (job_interests, notifications) and for real actions (accept / reject), while THIS store —
 * persisted in DataStore, scoped per user — owns whether each notification is read or cleared.
 *
 * Unread rule: `createdAt > baseline && id ∉ read && id ∉ cleared`. The `baseline` is stamped
 * the first time a user is configured on a build with this store, so the entire pre-existing
 * backlog is implicitly "read" and never inflates the badge — without a server migration.
 *
 * Trade-off (intentional, matches iOS): state lives on-device, so it does not sync across
 * devices and resets on reinstall.
 *
 * Reads ([isUnread]/[isCleared]) are synchronous against an in-memory mirror; writes persist
 * asynchronously and bump [revision] so observers (the feed + the bell badge) recompute.
 */
class NotificationLocalState(private val context: Context) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    /** Bumped on every mutation so observers re-render / recompute (iOS @Published revision). */
    private val _revision = MutableStateFlow(0)
    val revision: StateFlow<Int> = _revision.asStateFlow()

    @Volatile private var userId: String = ""
    @Volatile private var readIds: Set<String> = emptySet()
    @Volatile private var clearedIds: Set<String> = emptySet()
    // Until the per-user baseline is loaded, treat everything as read (MAX) so we never flash
    // the whole backlog as unread during the brief async load.
    @Volatile private var baseline: Long = Long.MAX_VALUE

    private fun readKey(uid: String) = stringSetPreferencesKey("notif.read.$uid")
    private fun clearedKey(uid: String) = stringSetPreferencesKey("notif.cleared.$uid")
    private fun baselineKey(uid: String) = longPreferencesKey("notif.baseline.$uid")

    /** Scope the store to a user and load their persisted state. Idempotent. */
    fun configure(uid: String) {
        if (uid.isEmpty() || uid == userId) return
        userId = uid
        baseline = Long.MAX_VALUE
        readIds = emptySet()
        clearedIds = emptySet()
        scope.launch { load(uid) }
    }

    private suspend fun load(uid: String) {
        runCatching {
            val prefs = context.notifDataStore.data.first()
            // Union with any mutations that landed before the async load finished.
            readIds = (prefs[readKey(uid)] ?: emptySet()) + readIds
            clearedIds = (prefs[clearedKey(uid)] ?: emptySet()) + clearedIds
            val stored = prefs[baselineKey(uid)]
            baseline = if (stored != null) {
                stored
            } else {
                val now = System.currentTimeMillis()
                context.notifDataStore.edit { it[baselineKey(uid)] = now }
                now
            }
        }
        bump()
    }

    // MARK: - Queries

    fun isCleared(id: String): Boolean = clearedIds.contains(id)

    fun isUnread(id: String, createdAtIso: String?): Boolean {
        if (clearedIds.contains(id) || readIds.contains(id)) return false
        val created = parseIsoMillis(createdAtIso) ?: return false
        return created > baseline
    }

    // MARK: - Mutations

    fun markRead(id: String) = markRead(listOf(id))

    fun markRead(ids: List<String>) {
        val newOnes = ids.filter { it !in readIds }
        if (newOnes.isEmpty()) return
        readIds = readIds + newOnes
        persistRead()
        bump()
    }

    fun clear(id: String) = clear(listOf(id))

    fun clear(ids: List<String>) {
        if (ids.isEmpty()) return
        clearedIds = clearedIds + ids
        readIds = readIds - ids.toSet()
        persistCleared()
        persistRead()
        bump()
    }

    // MARK: - Persistence

    private fun persistRead() {
        val uid = userId
        val snapshot = readIds
        scope.launch { runCatching { context.notifDataStore.edit { it[readKey(uid)] = snapshot } } }
    }

    private fun persistCleared() {
        val uid = userId
        val snapshot = clearedIds
        scope.launch { runCatching { context.notifDataStore.edit { it[clearedKey(uid)] = snapshot } } }
    }

    private fun bump() {
        _revision.value = _revision.value + 1
    }
}
